libopenage/game_control.cpp

```diff
@@ -41,13 +41,31 @@ void CreateMode::on_exit() {}
 
 void CreateMode::render() {}
 
+ActionModeSignals::ActionModeSignals(ActionMode *action_mode)
+	:
+	action_mode(action_mode) {
+}
+
+void ActionModeSignals::on_action(const std::string &action_name) {
+	Engine &engine = Engine::get();
+	input::action_t action = engine.get_action_manager().get(action_name);
+	input::InputContext *top_ctxt = &engine.get_input_manager().get_top_context();
+	if (top_ctxt == this->action_mode || top_ctxt == &this->action_mode->building_context ||
+	    top_ctxt == &this->action_mode->build_menu_context || top_ctxt == &this->action_mode->build_menu_mil_context) {
+		using namespace input;
```

Comment:
Found it, I stole it from [this commit](https://github.com/jazzpi/openage/commit/3fb34d74fc74bec4efee0c23be2e978fc70bbbb8#diff-3cd26de8330566a41cadb63f7de53cccR350).